core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala

```diff
@@ -591,22 +593,41 @@ class HttpGoogleServicesDAO(
-    retryWithRecoverWhen500orGoogleError(() => {
-      val list = blocking {
-        executeGoogleRequest(fetcher)
+
+    type Paginated[T] = (Option[T], Option[String])
+
+    def makeCall(pageToken: Option[String] = None): Future[Paginated[mutable.Buffer[BillingAccount]]] = {
+      retryWithRecoverWhen500orGoogleError(() => {
+        val result = executeGoogleListBillingAccountsRequest(credential, pageToken)
+        // option-wrap getBillingAccounts because it returns null for an empty list,
+        // and result.getNextPateToken = '' when there are no more pages.
+        (Option(result.getBillingAccounts.asScala), Option(if (result.getNextPageToken.isEmpty) null else result.getNextPageToken))
+      }) {
+        case gjre: GoogleJsonResponseException
+          if gjre.getStatusCode == StatusCodes.Forbidden.intValue &&
+            gjre.getDetails.getMessage == "Request had insufficient authentication scopes." =>
+          // This error message is purely informational. A client can determine which scopes it has
+          // been granted, so an insufficiently-scoped request would generally point to a programming error.
+          throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.Forbidden, BillingAccountScopes(billingScopes).toJson.toString))
       }
-      // option-wrap getBillingAccounts because it returns null for an empty list
-      Option(list.getBillingAccounts.asScala).map(_.toList).getOrElse(List.empty)
-    }) {
-      case gjre: GoogleJsonResponseException
-        if gjre.getStatusCode == StatusCodes.Forbidden.intValue &&
-          gjre.getDetails.getMessage == "Request had insufficient authentication scopes." =>
-        // This error message is purely informational. A client can determine which scopes it has
-        // been granted, so an insufficiently-scoped request would generally point to a programming
-        // error.
-        throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.Forbidden, BillingAccountScopes(billingScopes).toJson.toString))
     }
+
+    def recurse(acc: List[BillingAccount] = List.empty, pageToken: Option[String] = None): Future[List[BillingAccount]] = {
+      val result: Future[Paginated[mutable.Buffer[BillingAccount]]] = makeCall(pageToken)
+      result.flatMap {
+        case (Some(items), Some(nextPage)) => recurse(acc ::: items.toList, Some(nextPage))
+        case (Some(items), None) => Future.successful(acc ::: items.toList)
+        case _ => Future.successful(acc)
+      }
+    }
+    recurse()
+  }
+
+  protected def executeGoogleListBillingAccountsRequest(credential: Credential, pageToken: Option[String] = None)(implicit counters: GoogleCounters): ListBillingAccountsResponse = {
+    // To test with pagination, can call `setPageSize` at the end of the call below.
+    val fetcher = getCloudBillingManager(credential).billingAccounts().list()
+    pageToken.foreach(fetcher.setPageToken)
+    blocking { executeGoogleRequest(fetcher) }
```

Comment:
We do not consistently wrap `executeGoogleRequest` with blocking. I'm not sure how much it matters, as the code appears to behave the same way regardless.

Reply:
I think this lack of consistency could eventually lead to thread starvation since there may be other threads ready for execution while this one blocks. The real solution as I understand it is to create a dedicated dispatcher for blocking threads such as this one. But, out of scope for this PR I'd say.

Reply:
@aherbst-broad do you have any details beyond the documentation regarding what "blocking" does and how important it is to use it?

````
Used to designate a piece of code which potentially blocks, allowing the current BlockContext to adjust the runtime's behavior. 
Properly marking blocking code may improve performance or avoid deadlocks.
Blocking on an Awaitable should be done using Await.result instead of blocking.

Params:
body – A piece of code which contains potentially blocking or long running calls.
Throws:
scala.concurrent.CancellationException – if the computation was cancelled
InterruptedException – in the case that a wait within the blocking body was interrupted